src/Compilers/Extension/CompilerPackage.cs

```diff
@@ -90,8 +79,38 @@ private void WriteMSBuildFiles(string packagePath, string hiveName)
   </PropertyGroup>
 </Project>";
 
-                WriteMSBuildFile(targetsContent, $@"Microsoft.CSharp.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets");
-                WriteMSBuildFile(targetsContent, $@"Microsoft.VisualBasic.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets");
+            filesToWrite.Add(GetMSBuildRelativePath($@"Microsoft.CSharp.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets"), targetsContent);
+            filesToWrite.Add(GetMSBuildRelativePath($@"Microsoft.VisualBasic.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets"), targetsContent);
+
+            // First we want to ensure any Roslyn files with our hive name that we aren't writing -- this is probably
+            // leftovers from older extensions
+            var msbuildDirectory = new DirectoryInfo(GetMSBuildPath());
+            if (msbuildDirectory.Exists)
+            {
+                foreach (var file in msbuildDirectory.EnumerateFiles($"*Roslyn*{hiveName}*", SearchOption.AllDirectories))
+                {
+                    if (!filesToWrite.ContainsKey(file.FullName))
+                    {
+                        file.Delete();
+                    }
+                }
+            }
+
+            try
+            {
+                foreach (var fileAndContents in filesToWrite)
+                {
+                    var parentDirectory = new DirectoryInfo(Path.GetDirectoryName(fileAndContents.Key));
+                    parentDirectory.Create();
+
+                    // If we already know the file has the same contents, then we can skip
+                    if (File.Exists(fileAndContents.Key) && File.ReadAllText(fileAndContents.Key) == fileAndContents.Value)
+                    {
+                        continue;
+                    }
+
+                    File.WriteAllText(fileAndContents.Key, fileAndContents.Value);
```

Comment:
Nit: consider using an explicit encoding in all the file ops here. I often find this useful when doing content comparison. Reduces chances of enocding probs. Not a blocker.